lvmd/command/lvm.go

```diff
@@ -644,16 +644,102 @@ func (l *LogicalVolume) Snapshot(name string, cowSize uint64) (*LogicalVolume, e
 
 	var lvcreateArgs []string
 	if _, err := os.Stat("/run/systemd/system"); err != nil {
-		lvcreateArgs = []string{"-s", "-n", name, l.fullname}
+		lvcreateArgs = []string{"--type", "thin", "-s", "-n", name, l.fullname}
 	} else {
-		lvcreateArgs = []string{"-s", "-k", "n", "-n", name, l.fullname}
+		lvcreateArgs = []string{"--type", "thin", "-s", "-k", "n", "-n", name, l.fullname}
 	}
 	if err := CallLVM("lvcreate", lvcreateArgs...); err != nil {
 		return nil, err
 	}
 	return l.vg.FindVolume(name)
 }
 
+// Activate activates the logical volume for desired access.
+func (l *LogicalVolume) Activate(access string) error {
+	var lvchangeArgs []string
+	switch access {
+	case "ro":
+		lvchangeArgs = []string{"-p", "r", l.path}
+	case "rw":
+		lvchangeArgs = []string{"-a", "y", l.path}
+	default:
+		return fmt.Errorf("unknown access: %s for LogicalVolume %s", access, l.fullname)
+	}
+	if err := CallLVM("lvchange", lvchangeArgs...); err != nil {
+		return err
+	}
+	return nil
+}
+
+func Call(cmd string, args ...string) error {
+	c := wrapExecCommand(cmd, args...)
+	log.Info("invoking command", map[string]interface{}{
+		"cmd":  cmd,
+		"args": args,
+	})
+	c.Stderr = os.Stderr
+	return c.Run()
+}
+
+func (l *LogicalVolume) UpdateFSUUID(volumeMode, fsType string) error {
```

Comment:
Yes, it is for XFS based file systems
- XFS filesystem has UUID which are unique identifiers of the filesystem. Two file systems with the same UUID can not be mounted on the same system.

- If the backing storage of the filesystem is cloned or replicated, the metadata that contains UUID information for the filesystem will be identical on both filesystems. This means that the LVM snapshots will have the same UUID as the original filesystem.

- Since for restored PVC we use thin-snapshots of the source logical volume, using the same UUID with fail to mount the filesystem with error 
  ```
  # dmesg | tail
  Oct 16 03:29:22 hostname kernel: XFS (sdc1): Filesystem has duplicate UUID c4e2eef9-c744-4c64-8ebc-044373f84ddc - can't mount
  ```

- To avoid this issue, we need to generate a new UUID for the XFS filesystem using `xfs_admin`.

- So before nodepublish, we generate a new UUID for XFS filesystem, to avoid the XFS filesystem UUID clash
 See: https://github.com/topolvm/topolvm/pull/463/commits/bec36d39733f54762733ea7a5faf372659dd7b00


For more information, See https://access.redhat.com/solutions/5494781